src/lib/utils/segment_tree.hpp

```diff
@@ -0,0 +1,137 @@
+      : leaf_count(std::bit_ceil(leaf_values.size())),
+        combine(std::move(init_combine)),
+        make_neutral_element(std::move(init_make_neutral_element)),
+        tree(2 * leaf_count + 1, make_neutral_element()) {
+    const auto first_leaf_index = leaf_count;
+    std::ranges::copy(leaf_values, &tree[first_leaf_index]);
+
+    for (auto node = first_leaf_index - 1; root <= node; --node) {
+      tree[node] = combine(tree[left_child(node)], tree[right_child(node)]);
+    }
+  }
+
+  // Returns the aggregate of the leaf values in the queried range.
+  T range_query(QueryRange range) const {
+    // The algorithm works like this: There are two pointers, left_node and right_node, that span the range that still
+    // needs to be included in the query aggregate. Initially, they start the corresponding leaf node values. As with
+    // the query interface, the left_node is inclusive and the right_node is exclusive (that is, right_node could point
+    // outside of the tree array).
+    // In each iteration, the pointers update one level towards the root of the tree. Whenever the next parent of one
+    // the pointers would not be fully contained in `range`, the current nodes value is `combine`d into the aggregate
+    // result and the pointer moves inward. After moving inward, the new nodes parent must be fully included in the
+    // range again. A node must be included exactly if its index is odd, because each level (except for the root) starts
+    // with an even index and the leftmost node is always a left child.
+    //
+    // This process continues until both pointers point to the same node. This final node is explicitly not included in
+    // the query aggregate, because by definition, the right pointer is exclusive, so the interval of the two pointers
+    // is now empty.
+    //
+    // Tiny example: sum query for the range [1, 4):
+    // (read algorithm steps from bottom to top; l and r are short for left_node and right_node respectively)
+    //
+    //                  6        l and r are both after 6 (index 2) ; query is done
+    //                1   5      l is on 5, r is after 5 (index 4)  ; 5 is included from l in the aggregate
+    //               0 1 2 3     l is on 1, r is after 3 (index 8)  ; 1 is included from l in the aggregate
+    //
+    // Leaf index:   4 5 6 7
+
+    DebugAssert(range.start <= range.end, "Got malformed Range in SegmentTree::range_query.");
+    DebugAssert(range.end <= leaf_count, "Query range out of bounds.");
+
+    // Note: To avoid the commutativity assumption, two separate aggregates are needed.
+    auto agg = make_neutral_element();
+    auto left_node = leaf_count + range.start;
+    auto right_node = leaf_count + range.end;
+    while (left_node < right_node) {
+      if (left_node & 1u)
+        agg = combine(tree[left_node++], agg);
+      if (right_node & 1u)
+        agg = combine(agg, tree[--right_node]);
```

Comment:
Maybe add a comment that you are a right child when this case happens.

Reply:
We finally decided that we will keep the algorithm brief in favor of having the extensive comment above that already explains this